File: src/MixNetFx/FacessoData/SQLServer/SPAccess_TimeLog.vb
Imports Activedev
Imports Facesso.Data
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    Friend Sub TimeLog_DeleteItemFromDatabase(ByVal WorkGroup As WorkGroupInfo, ByVal EmpTimeLogItem As EmployeeTimeLogInfoItem)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection

            Dim locCmd As New SqlCommand("TimeLog_DeleteItem", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = WorkGroup.IDSubsidiary
                .Add("@IDTimeLog", SqlDbType.BigInt).Value = EmpTimeLogItem.IDTimeLog
            End With
            locCmd.ExecuteNonQuery()
        End Using
    End Sub

    Friend Sub GetEmployeeTimeLog(ByVal TimeLogItems As EmployeeTimeLogInfo)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()

        TimeLogItems.Clear()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection

            Dim locCmd As New SqlCommand("TimeLog_GetLogItemsForShiftDate", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDSubsidiary
                .Add("@IDWorkGroup", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDWorkGroup
                .Add("@ProductionDate", SqlDbType.DateTime).Value = TimeLogItems.ProductionDate
                .Add("@Shift", SqlDbType.TinyInt).Value = TimeLogItems.Shift
            End With
            Dim locDR As SqlDataReader = locCmd.ExecuteReader()

            Do While locDR.Read
                Dim locTimeLogItem As New EmployeeTimeLogInfoItem(locDR, True)
                TimeLogItems.Add(locTimeLogItem)
            Loop
        End Using
    End Sub

    Friend Sub GetEmployeeTimelog(ByVal Employee As EmployeeInfo, ByVal Startdate As Date, ByVal Enddate As Date, ByVal TimeLogItems As EmployeeTimeLogInfo)
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()

        TimeLogItems.Clear()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection

            Dim locCmd As New SqlCommand("TimeLog_GetLogItemsForRange", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = Employee.IDSubsidiary
                .Add("@IDEmployee", SqlDbType.Int).Value = Employee.IDEmployee
                .Add("@StartDate", SqlDbType.DateTime).Value = Startdate
                .Add("@EndDate", SqlDbType.DateTime).Value = Enddate
            End With
            Dim locDR As SqlDataReader = locCmd.ExecuteReader()

            Do While locDR.Read
                Dim locTimeLogItem As New EmployeeTimeLogInfoItem(locDR, True)
                TimeLogItems.Add(locTimeLogItem)
            Loop
        End Using
    End Sub

    Friend Sub TimeLog_GetOverlappingLogItems(ByVal EmployeeInfo As EmployeeInfo, _
                                             ByVal ShiftStart As Date, _
                                             ByVal ShiftEnd As Date, _
                                             ByVal OverlapsInfo As OverlapsInfo, _
                                             ByVal ExcludeIDTimelog As ADDBNullable(Of Long))


        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection

            Dim locCmd As New SqlCommand("TimeLog_GetOverlappingLogItems", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = EmployeeInfo.IDSubsidiary
                .Add("@IDEmployee", SqlDbType.Int).Value = EmployeeInfo.IDEmployee
                .Add("@ShiftStart", SqlDbType.DateTime).Value = ShiftStart
                .Add("@ShiftEnd", SqlDbType.DateTime).Value = ShiftEnd
                .Add("@ExcludeIDTimeLog", SqlDbType.BigInt).Value = ExcludeIDTimelog.Value
            End With
            Dim locDR As SqlDataReader = locCmd.ExecuteReader()

            Do While locDR.Read
                Dim locItem As New OverlapsInfoItem(EmployeeInfo, locDR)
                OverlapsInfo.Add(locItem)
            Loop
        End Using
    End Sub

    Friend Function TimeLog_AddEditEmployeeTimeLogItems(ByVal TimeLogItems As EmployeeTimeLogInfo, ByVal IDUser As Integer, ByVal ReturnResultSet As Boolean) As EmployeeTimeLogInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection
            Dim locCmd As SqlCommand
            Dim locTicket As DateTime = Date.Now

            For Each locItem As EmployeeTimeLogInfoItem In TimeLogItems
                locCmd = New SqlCommand("TimeLog_AddItemsForAddEdit", locConnection)
                locCmd.CommandType = CommandType.StoredProcedure
                If locItem.Deleted And locItem.IDTimeLog < 1 Then Continue For

                With locCmd.Parameters
                    .Add("@IDSubsidiary", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDSubsidiary
                    .Add("@IDTimeLog", SqlDbType.BigInt).Value = locItem.IDTimeLog
                    .Add("@IDUser", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDUser
                    .Add("@IDWorkGroup", SqlDbType.Int).Value = locItem.IDWorkGroup
                    .Add("@IDEmployee", SqlDbType.Int).Value = locItem.EmployeeInfo.IDEmployee
                    .Add("@ProductionDate", SqlDbType.DateTime).Value = locItem.ProductionDate
                    .Add("@Shift", SqlDbType.TinyInt).Value = locItem.Shift
                    .Add("@ShiftStart", SqlDbType.DateTime).Value = locItem.ShiftStart
                    .Add("@ShiftEnd", SqlDbType.DateTime).Value = locItem.ShiftEnd
                    .Add("@WorkBreak", SqlDbType.Int).Value = locItem.WorkBreak
                    .Add("@DownTime", SqlDbType.Int).Value = locItem.DownTime
                    .Add("@Handicap", SqlDbType.Int).Value = locItem.Handicap
                    .Add("@InsertedByInterface", SqlDbType.Bit).Value = locItem.InsertedByInterface
                    .Add("@ManuallyEdited", SqlDbType.Bit).Value = locItem.ManuallyEdited
                    .Add("@LastEditedByIDUser", SqlDbType.Int).Value = IDUser
                    .Add("@Deleted", SqlDbType.Bit).Value = locItem.Deleted
                    .Add("@Ticket", SqlDbType.DateTime).Value = locTicket
                End With
                locCmd.ExecuteNonQuery()
            Next

            locCmd = New SqlCommand("TimeLog_HandleAddEdit", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDSubsidiary
                .Add("@IDUser", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDUser
                .Add("@Ticket", SqlDbType.DateTime).Value = locTicket
            End With
            locCmd.CommandTimeout = 300
            locCmd.ExecuteNonQuery()

            If ReturnResultSet Then
                locCmd = New SqlCommand("TimeLog_GetLogItemsForShiftDate", locConnection)
                locCmd.CommandType = CommandType.StoredProcedure
                TimeLogItems.Clear()

                With locCmd.Parameters
                    .Add("@IDSubsidiary", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDSubsidiary
                    .Add("@IDWorkGroup", SqlDbType.Int).Value = TimeLogItems.WorkGroup.IDWorkGroup
                    .Add("@ProductionDate", SqlDbType.DateTime).Value = TimeLogItems.ProductionDate
                    .Add("@Shift", SqlDbType.TinyInt).Value = TimeLogItems.Shift
                End With
                Dim locDR As SqlDataReader = locCmd.ExecuteReader()

                Do While locDR.Read
                    Dim locTimeLogItem As New EmployeeTimeLogInfoItem(locDR, True)
                    TimeLogItems.Add(locTimeLogItem)
                Loop
                Return TimeLogItems
            Else
                Return Nothing
            End If
        End Using
    End Function

    Friend Sub TimeLog_GetEmployeeResult(ByVal IDSubsidiary As Integer, ByVal IDUser As Integer, _
                    ByVal Ticket As Date, ByVal Employee As EmployeeInfo, ByVal TimeLogItems As EmployeeTimeLogInfo)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        TimeLogItems.Clear()

        If locConnection Is Nothing Then
            Dim locUp As New FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", Nothing)
            Throw locUp
        End If

        Using locConnection

            Dim locCmd As New SqlCommand("TimeLog_Analysis_GetEmployeeResult", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDUser", SqlDbType.Int).Value = IDUser
                .Add("@Ticket", SqlDbType.DateTime).Value = Ticket
                .Add("@IDEmployee", SqlDbType.Int).Value = Employee.IDEmployee
            End With
            Dim locDR As SqlDataReader = locCmd.ExecuteReader()

            Do While locDR.Read
                Dim locTimeLogItem As New EmployeeTimeLogInfoItem(locDR, Employee)
                TimeLogItems.Add(locTimeLogItem)
            Loop
        End Using
    End Sub
End Class

File: src/MixNetFx/Facesso.Data/SQLServer/SPAccess_TimeLog.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;
using Facesso;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        internal void TimeLog_DeleteItemFromDatabase(WorkGroupInfo workGroup, EmployeeTimeLogInfoItem empTimeLogItem)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                var locCmd = new SqlCommand("TimeLog_DeleteItem", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = workGroup.IDSubsidiary;
                locCmd.Parameters.Add("@IDTimeLog", SqlDbType.BigInt).Value = empTimeLogItem.IDTimeLog;
                locCmd.ExecuteNonQuery();
            }
        }

        internal void GetEmployeeTimeLog(EmployeeTimeLogInfo timeLogItems)
        {
            timeLogItems.Clear();
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                var locCmd = new SqlCommand("TimeLog_GetLogItemsForShiftDate", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDSubsidiary;
                locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDWorkGroup;
                locCmd.Parameters.Add("@ProductionDate", SqlDbType.DateTime).Value = timeLogItems.ProductionDate;
                locCmd.Parameters.Add("@Shift", SqlDbType.TinyInt).Value = timeLogItems.Shift;
                SqlDataReader locDR = locCmd.ExecuteReader();
                while (locDR.Read())
                {
                    var locTimeLogItem = new EmployeeTimeLogInfoItem(locDR, true);
                    timeLogItems.Add(locTimeLogItem);
                }
            }
        }

        internal void GetEmployeeTimeLog(EmployeeInfo employee, DateTime startdate, DateTime enddate, EmployeeTimeLogInfo timeLogItems)
        {
            timeLogItems.Clear();
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                var locCmd = new SqlCommand("TimeLog_GetLogItemsForRange", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = employee.IDSubsidiary;
                locCmd.Parameters.Add("@IDEmployee", SqlDbType.Int).Value = employee.IDEmployee;
                locCmd.Parameters.Add("@StartDate", SqlDbType.DateTime).Value = startdate;
                locCmd.Parameters.Add("@EndDate", SqlDbType.DateTime).Value = enddate;
                SqlDataReader locDR = locCmd.ExecuteReader();
                while (locDR.Read())
                {
                    var locTimeLogItem = new EmployeeTimeLogInfoItem(locDR, true);
                    timeLogItems.Add(locTimeLogItem);
                }
            }
        }

        internal void TimeLog_GetOverlappingLogItems(EmployeeInfo employeeInfo, DateTime shiftStart, DateTime shiftEnd,
            OverlapsInfo overlapsInfo, ADDBNullable<long> excludeIDTimelog)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                var locCmd = new SqlCommand("TimeLog_GetOverlappingLogItems", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = employeeInfo.IDSubsidiary;
                locCmd.Parameters.Add("@IDEmployee", SqlDbType.Int).Value = employeeInfo.IDEmployee;
                locCmd.Parameters.Add("@ShiftStart", SqlDbType.DateTime).Value = shiftStart;
                locCmd.Parameters.Add("@ShiftEnd", SqlDbType.DateTime).Value = shiftEnd;
                locCmd.Parameters.Add("@ExcludeIDTimeLog", SqlDbType.BigInt).Value = excludeIDTimelog.Value;
                SqlDataReader locDR = locCmd.ExecuteReader();
                while (locDR.Read())
                {
                    var locItem = new OverlapsInfoItem(employeeInfo, locDR);
                    overlapsInfo.Add(locItem);
                }
            }
        }

        internal EmployeeTimeLogInfo TimeLog_AddEditEmployeeTimeLogItems(EmployeeTimeLogInfo timeLogItems, int idUser, bool returnResultSet)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                SqlCommand locCmd;
                DateTime locTicket = DateTime.Now;

                foreach (EmployeeTimeLogInfoItem locItem in timeLogItems)
                {
                    if (locItem.Deleted && locItem.IDTimeLog < 1) continue;

                    locCmd = new SqlCommand("TimeLog_AddItemsForAddEdit", locConnection);
                    locCmd.CommandType = CommandType.StoredProcedure;
                    locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDSubsidiary;
                    locCmd.Parameters.Add("@IDTimeLog", SqlDbType.BigInt).Value = locItem.IDTimeLog;
                    locCmd.Parameters.Add("@IDUser", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDUser;
                    locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = locItem.IDWorkGroup;
                    locCmd.Parameters.Add("@IDEmployee", SqlDbType.Int).Value = locItem.EmployeeInfo.IDEmployee;
                    locCmd.Parameters.Add("@ProductionDate", SqlDbType.DateTime).Value = locItem.ProductionDate;
                    locCmd.Parameters.Add("@Shift", SqlDbType.TinyInt).Value = locItem.Shift;
                    locCmd.Parameters.Add("@ShiftStart", SqlDbType.DateTime).Value = locItem.ShiftStart;
                    locCmd.Parameters.Add("@ShiftEnd", SqlDbType.DateTime).Value = locItem.ShiftEnd;
                    locCmd.Parameters.Add("@WorkBreak", SqlDbType.Int).Value = locItem.WorkBreak;
                    locCmd.Parameters.Add("@DownTime", SqlDbType.Int).Value = locItem.DownTime;
                    locCmd.Parameters.Add("@Handicap", SqlDbType.Int).Value = locItem.Handicap;
                    locCmd.Parameters.Add("@InsertedByInterface", SqlDbType.Bit).Value = locItem.InsertedByInterface;
                    locCmd.Parameters.Add("@ManuallyEdited", SqlDbType.Bit).Value = locItem.ManuallyEdited;
                    locCmd.Parameters.Add("@LastEditedByIDUser", SqlDbType.Int).Value = idUser;
                    locCmd.Parameters.Add("@Deleted", SqlDbType.Bit).Value = locItem.Deleted;
                    locCmd.Parameters.Add("@Ticket", SqlDbType.DateTime).Value = locTicket;
                    locCmd.ExecuteNonQuery();
                }

                locCmd = new SqlCommand("TimeLog_HandleAddEdit", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDSubsidiary;
                locCmd.Parameters.Add("@IDUser", SqlDbType.Int).Value = FacessoGeneric.LoginInfo.IDUser;
                locCmd.Parameters.Add("@Ticket", SqlDbType.DateTime).Value = locTicket;
                locCmd.CommandTimeout = 300;
                locCmd.ExecuteNonQuery();

                if (returnResultSet)
                {
                    locCmd = new SqlCommand("TimeLog_GetLogItemsForShiftDate", locConnection);
                    locCmd.CommandType = CommandType.StoredProcedure;
                    timeLogItems.Clear();
                    locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDSubsidiary;
                    locCmd.Parameters.Add("@IDWorkGroup", SqlDbType.Int).Value = timeLogItems.WorkGroup.IDWorkGroup;
                    locCmd.Parameters.Add("@ProductionDate", SqlDbType.DateTime).Value = timeLogItems.ProductionDate;
                    locCmd.Parameters.Add("@Shift", SqlDbType.TinyInt).Value = timeLogItems.Shift;
                    SqlDataReader locDR = locCmd.ExecuteReader();
                    while (locDR.Read())
                    {
                        var locTimeLogItem = new EmployeeTimeLogInfoItem(locDR, true);
                        timeLogItems.Add(locTimeLogItem);
                    }
                    return timeLogItems;
                }
                return null;
            }
        }

        internal void TimeLog_GetEmployeeResult(int idSubsidiary, int idUser, DateTime ticket,
            EmployeeInfo employee, EmployeeTimeLogInfo timeLogItems)
        {
            timeLogItems.Clear();
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null)
                throw new FacessoSqlDbException("Could not reach Facesso-Database while running StoredProcedure GetProductionItems", null);

            using (locConnection)
            {
                var locCmd = new SqlCommand("TimeLog_Analysis_GetEmployeeResult", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDUser", SqlDbType.Int).Value = idUser;
                locCmd.Parameters.Add("@Ticket", SqlDbType.DateTime).Value = ticket;
                locCmd.Parameters.Add("@IDEmployee", SqlDbType.Int).Value = employee.IDEmployee;
                SqlDataReader locDR = locCmd.ExecuteReader();
                while (locDR.Read())
                {
                    var locTimeLogItem = new EmployeeTimeLogInfoItem(locDR, employee);
                    timeLogItems.Add(locTimeLogItem);
                }
            }
        }
    }
}
